parallaxinc/Parallax-IDE
# Undo buffers

#### What's this PR do?

Adds document buffers to Parallax IDE to allow for file specific undo/redo.
#### Where should the reviewer start?

Pull this branch.
npm run build
#### How should this be manually tested?

Create a few test files. Add text to each. Click back and forth between files to test undo. Undo should not leave the active file.

Try to create new files and delete files. Be sure undo/redo is behaving as you would expect.
#### Any background context you want to provide?

A few interesting cases needed to be dealt with in order to integrate CodeMirror document buffers with the existing file system. The buffer system itself is relatively straight forward. Each time the user navigates to a file, a new buffer is generated if it does not already exist and swapped into the editor. If it does exist, the document is pulled from the buffer in memory. However at all times, the workspace is updating content in the editor regardless of what buffer is present. This created some timing challenges. I needed to have the buffer document present in the background but ignore changes that would impact the undo history. This is what the allowUpdate logic accounts for in the editor index.js file.

Another timing issue presented itself on initial startup of the editor. Our current implementation has the first file loaded into the workspace before the CodeMirror instance actually exists. I had to code a special case for that in the file store.

Much of the case off coding could be avoided if files were loaded directly into a CodeMirror document. However that would involve a rework of the background file interaction system. That  might be worth the effort to lower complexity of this buffer implementation and future work that deals with projects/file system behavior.
#### What are the relevant tickets?

Closes #69
